File: src/main/java/by/masarnovsky/releasedemon/client/LastfmClient.java
package by.masarnovsky.releasedemon.client;

import by.masarnovsky.releasedemon.lastfm.LastfmUserLibraryResponse;
import org.springframework.cloud.openfeign.FeignClient;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;

@FeignClient(value = "lastfmclient", url = "http://ws.audioscrobbler.com/2.0")
public interface LastfmClient {

    @RequestMapping(method = RequestMethod.GET, value = "/?method=library.getartists&format=json&")
    LastfmUserLibraryResponse retrievePageWithArtists(@RequestParam("api_key") String apiKey, @RequestParam("user") String user, @RequestParam("page") Integer page);
}

File: src/main/kotlin/by/masarnovsky/releasedemon/backend/external/service/LastfmClient.kt
package by.masarnovsky.releasedemon.backend.external.service

import by.masarnovsky.releasedemon.backend.external.dto.LastfmUserLibraryResponse
import com.github.kittinunf.fuel.httpGet
import com.github.kittinunf.fuel.jackson.responseObject
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

@Service
class LastfmClient {

    @Value("\${last.fm.api.key}")
    lateinit var apiKey: String

    fun retrievePageWithArtists(username: String, page: Int): LastfmUserLibraryResponse? {
        val lastfmUrl =
            "https://ws.audioscrobbler.com/2.0/?method=library.getartists&format=json&api_key=$apiKey&user=$username&page=$page"

        val (_, _, result) = lastfmUrl
            .httpGet()
            .responseObject<LastfmUserLibraryResponse>()

        return result.component1()
    }
}